### PR TITLE
[WIP] conda build scripts for Windows

### PR DESCRIPTION
**Pull request by yf225**
This PR is based on @peterjc123's initial scripts with some modifications. It's verified on the following matrix:

Backend:
- CPU
- CUDA 8.0
- CUDA 9.0
- CUDA 9.1

Python version:
- Python 2.7
- Python 3.5
- Python 3.6

I have yet to test the edge cases outlined in README todo, will do it and follow up in this PR.

### CODE REVIEW

File: conda/build_pytorch.sh [hunk 1]
```diff
@@ -26,37 +26,67 @@ if [[ "$OSTYPE" == "darwin"* ]]; then
...
+        time conda build -c $ANACONDA_USER --no-anaconda-upload --python 3.5 pytorch-$BUILD_VERSION
+        time conda build -c $ANACONDA_USER --no-anaconda-upload --python 3.6 pytorch-$BUILD_VERSION
+    elif [[ "$1" == "90" ]]; then
+        export CUDA_VERSION="9.0"
+        export CUDNN_VERSION="7.0"
+        time conda build -c $ANACONDA_USER --no-anaconda-upload --python 2.7 pytorch-cuda90-$BUILD_VERSION
+        time conda build -c $ANACONDA_USER --no-anaconda-upload --python 3.5 pytorch-cuda90-$BUILD_VERSION
+        time conda build -c $ANACONDA_USER --no-anaconda-upload --python 3.6 pytorch-cuda90-$BUILD_VERSION
+    elif [[ "$1" == "91" ]]; then
+        export CUDA_VERSION="9.1"
```
Comment (peterjc123) [inline conda/build_pytorch.sh]: We can reuse . ./switch_cuda_version.sh 9.0 like the following. You can take a look at `C:\builder\switch_cuda_version.sh` on the dev machine.

File: conda/build_pytorch.sh [hunk 2]
```diff
@@ -26,37 +26,67 @@ if [[ "$OSTYPE" == "darwin"* ]]; then
...
+    conda install -y conda-build=3.1.1 # The newest version will give "No module named torch" error
+    conda build vs2017
+    if [[ "$1" == "80" ]]; then
+        export CUDA_VERSION="8.0"
+        export CUDNN_VERSION="7.0"
+        time conda build -c $ANACONDA_USER --no-anaconda-upload --python 2.7 pytorch-$BUILD_VERSION
+        time conda build -c $ANACONDA_USER --no-anaconda-upload --python 3.5 pytorch-$BUILD_VERSION
+        time conda build -c $ANACONDA_USER --no-anaconda-upload --python 3.6 pytorch-$BUILD_VERSION
+    elif [[ "$1" == "90" ]]; then
+        export CUDA_VERSION="9.0"
```
Comment (peterjc123) [inline conda/build_pytorch.sh]: Same with below.

File: conda/build_pytorch.sh [hunk 3]
```diff
@@ -26,37 +26,67 @@ if [[ "$OSTYPE" == "darwin"* ]]; then
     time conda build -c $ANACONDA_USER --no-anaconda-upload --python 2.7 pytorch-$BUILD_VERSION
     time conda build -c $ANACONDA_USER --no-anaconda-upload --python 3.5 pytorch-$BUILD_VERSION
     time conda build -c $ANACONDA_USER --no-anaconda-upload --python 3.6 pytorch-$BUILD_VERSION
+elif [[ "$OSTYPE" == "msys" ]]; then
+    conda uninstall -y conda-build
+    conda install -y conda-build=3.1.1 # The newest version will give "No module named torch" error
+    conda build vs2017
+    if [[ "$1" == "80" ]]; then
+        export CUDA_VERSION="8.0"
```
Comment (peterjc123) [inline conda/build_pytorch.sh]: Same with below.

File: conda/build_pytorch.sh [hunk 4]
```diff
@@ -26,37 +26,67 @@ if [[ "$OSTYPE" == "darwin"* ]]; then
     time conda build -c $ANACONDA_USER --no-anaconda-upload --python 2.7 pytorch-$BUILD_VERSION
     time conda build -c $ANACONDA_USER --no-anaconda-upload --python 3.5 pytorch-$BUILD_VERSION
     time conda build -c $ANACONDA_USER --no-anaconda-upload --python 3.6 pytorch-$BUILD_VERSION
+elif [[ "$OSTYPE" == "msys" ]]; then
+    conda uninstall -y conda-build
+    conda install -y conda-build=3.1.1 # The newest version will give "No module named torch" error
```
Comment (peterjc123) [inline conda/build_pytorch.sh]: It's not the conda-build's fault. Actually, it introduced a new method for writing meta.yaml. You can take a look at `C:\builder\conda\pytorch-cpu-0.3.1`. We should use newer version of conda-build.

File: conda/pytorch-0.3.1/bld.bat [hunk 5]
```diff
@@ -0,0 +1,49 @@
+@echo off
+
+set CUDA_VERSION=80
+set CUDNN_VERSION=7
```
Comment (peterjc123) [inline conda/pytorch-0.3.1/bld.bat]: Not necessary I think.

File: conda/pytorch-0.3.1/meta.yaml [hunk 6]
```diff
@@ -8,21 +8,24 @@ source:
...
     - mkl
-    - magma-cuda80 # [not osx]
+    - typing
+    - magma-cuda80 # [not osx and not win]
+    - ninja # [win]
+    - {{ compiler('c') }} # [win]
 
   run:
-    - python
+    - python {{PY_VER}}*,>=2.7,<3.3|>=3.4
```
Comment (peterjc123) [inline conda/pytorch-0.3.1/meta.yaml]: Why is this change needed?

File: conda/pytorch-cpu-0.3.1/meta.yaml [hunk 7]
```diff
@@ -8,15 +8,18 @@ source:
 
 requirements:
   build:
-    - python
+    - python {{PY_VER}}*,>=2.7,<3.3|>=3.4
```
Comment (peterjc123) [inline conda/pytorch-cpu-0.3.1/meta.yaml]: Same with above.

File: conda/pytorch-cpu-0.3.1/meta.yaml [hunk 8]
```diff
@@ -8,15 +8,18 @@ source:
...
     - pyyaml
     - cffi
     - mkl
+    - typing
+    - ninja # [win]
+    - {{ compiler('c') }} # [win]
 
   run:
-    - python
+    - python {{PY_VER}}*,>=2.7,<3.3|>=3.4
```
Comment (peterjc123) [inline conda/pytorch-cpu-0.3.1/meta.yaml]: Same with above.

File: conda/pytorch-cuda90-0.3.1/bld.bat [hunk 9]
```diff
@@ -0,0 +1,44 @@
+@echo off
+
+set CUDA_VERSION=90
+set CUDNN_VERSION=7
```
Comment (peterjc123) [inline conda/pytorch-cuda90-0.3.1/bld.bat]: Not necessary.

File: conda/pytorch-cuda90-0.3.1/meta.yaml [hunk 10]
```diff
@@ -8,16 +8,19 @@ source:
 
 requirements:
   build:
-    - python
+    - python {{PY_VER}}*,>=2.7,<3.3|>=3.4
```
Comment (peterjc123) [inline conda/pytorch-cuda90-0.3.1/meta.yaml]: Same with above.

File: conda/pytorch-cuda90-0.3.1/meta.yaml [hunk 11]
```diff
@@ -8,16 +8,19 @@ source:
...
     - mkl
-    - magma-cuda90 # [not osx]
+    - typing
+    - magma-cuda90 # [not osx and not win]
+    - ninja # [win]
+    - {{ compiler('c') }} # [win]
 
   run:
-    - python
+    - python {{PY_VER}}*,>=2.7,<3.3|>=3.4
```
Comment (peterjc123) [inline conda/pytorch-cuda90-0.3.1/meta.yaml]: Same with above.

File: conda/pytorch-cuda91-0.3.1/bld.bat [hunk 12]
```diff
@@ -0,0 +1,44 @@
+@echo off
+
+set CUDA_VERSION=91
+set CUDNN_VERSION=7
```
Comment (peterjc123) [inline conda/pytorch-cuda91-0.3.1/bld.bat]: Not necessary.

File: conda/pytorch-cuda91-0.3.1/meta.yaml [hunk 13]
```diff
@@ -8,16 +8,19 @@ source:
 
 requirements:
   build:
-    - python
+    - python {{PY_VER}}*,>=2.7,<3.3|>=3.4
```
Comment (peterjc123) [inline conda/pytorch-cuda91-0.3.1/meta.yaml]: Same with above.

File: conda/pytorch-cuda91-0.3.1/meta.yaml [hunk 14]
```diff
@@ -8,16 +8,19 @@ source:
...
     - mkl
-    - magma-cuda91 # [not osx]
+    - typing
+    - magma-cuda91 # [not osx and not win]
+    - ninja # [win]
+    - {{ compiler('c') }} # [win]
 
   run:
-    - python
+    - python {{PY_VER}}*,>=2.7,<3.3|>=3.4
```
Comment (peterjc123) [inline conda/pytorch-cuda91-0.3.1/meta.yaml]: Same with above.

File: conda/build_pytorch.sh [hunk 15]
```diff
@@ -26,37 +26,67 @@ if [[ "$OSTYPE" == "darwin"* ]]; then
...
+    conda build vs2017
+    if [[ "$1" == "80" ]]; then
+        export CUDA_VERSION="8.0"
+        export CUDNN_VERSION="7.0"
+        time conda build -c $ANACONDA_USER --no-anaconda-upload --python 2.7 pytorch-$BUILD_VERSION
```
Comment (peterjc123) [inline conda/build_pytorch.sh]: Does the package runs well under 2.7? Do all tests pass?

File: conda/build_pytorch.sh (hunk 4 above)
Comment (peterjc123) [reply conda/build_pytorch.sh]: [Here](https://github.com/conda/conda-build/issues/2698#event-1488393700) is related issue I made.

**Comment by peterjc123** [comment]
I think it looks good. But here are some minor points that may need reconsideration. 
Besides, there is an issue that needs discussion, that is, since the code has to be compiled on WS 2012 and 2016, how are we going to deal with the packaging? There're two possible ways.
1. Repackage the initial packages and add a if logic that runs once to delete the unused files.
2. Make separate packages for both platforms. In detail:
2a. Name them differently. Like pytorch for WS2016/Win 10 and pytorch_legacy for WS2012/Win7
2b. Use a tracking feature, just like what `cuda90` do.
Of all these methods, 1 is the most ideal one. But the conda package may be hard to repackage. Compared to 2b, 2a is better because it can be done using environment variables. In 2b, we have to copy the scripts and meta files, which is not so good.
@soumith What do you think?

File: conda/pytorch-cuda90-0.3.1/meta.yaml [hunk 16]
```diff
@@ -8,16 +8,19 @@ source:
...
     - numpy 1.11.*
     - setuptools
     - pyyaml
     - cffi
     - mkl
-    - magma-cuda90 # [not osx]
+    - typing
+    - magma-cuda90 # [not osx and not win]
+    - ninja # [win]
```
Comment (peterjc123) [inline conda/pytorch-cuda90-0.3.1/meta.yaml]: Why is ninja limited to windows?

File: conda/pytorch-cuda91-0.3.1/bld.bat [hunk 17]
```diff
@@ -0,0 +1,44 @@
+@echo off
+
+set CUDA_VERSION=91
+set CUDNN_VERSION=7
+set CUDA_PATH_V9_1=C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v9.1
+set CUDA_PATH=%CUDA_PATH_V9_1%
+
+set CUDA_BIN_PATH=%CUDA_PATH%\bin
+set PATH=%CUDA_BIN_PATH%;C:\Program Files\CMake\bin;%PATH%
+set TORCH_CUDA_ARCH_LIST=3.5;5.2+PTX;6.0;6.1
```
Comment (peterjc123) [inline conda/pytorch-cuda91-0.3.1/bld.bat]: Why is 7.0 ignored?

File: conda/pytorch-cuda90-0.3.1/bld.bat [hunk 18]
```diff
@@ -0,0 +1,44 @@
+@echo off
+
+set CUDA_VERSION=90
+set CUDNN_VERSION=7
+set CUDA_PATH_V9_0=C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v9.0
+set CUDA_PATH=%CUDA_PATH_V9_0%
+
+set CUDA_BIN_PATH=%CUDA_PATH%\bin
+set PATH=%CUDA_BIN_PATH%;C:\Program Files\CMake\bin;%PATH%
+set TORCH_CUDA_ARCH_LIST=3.5;5.2+PTX;6.0;6.1
```
Comment (peterjc123) [inline conda/pytorch-cuda90-0.3.1/bld.bat]: Why is 7.0 ignored?

File: conda/pytorch-cuda90-0.3.1/meta.yaml (hunk 16 above)
Comment (yf225) [reply conda/pytorch-cuda90-0.3.1/meta.yaml]: ninja isn't used in Linux builds currently

File: conda/pytorch-cuda91-0.3.1/meta.yaml [hunk 19]
```diff
@@ -8,13 +8,19 @@ source:
 
 requirements:
   build:
+    - cmake
+    - setuptools
+    - ninja # [win]
```
Comment (peterjc123) [inline conda/pytorch-cuda91-0.3.1/meta.yaml]: Remove the [win] label is fine.

File: conda/pytorch-cuda90-0.3.1/meta.yaml [hunk 20]
```diff
@@ -8,13 +8,19 @@ source:
 
 requirements:
   build:
+    - cmake
+    - setuptools
+    - ninja # [win]
```
Comment (peterjc123) [inline conda/pytorch-cuda90-0.3.1/meta.yaml]: Same here.

File: conda/pytorch-cpu-0.3.1/meta.yaml [hunk 21]
```diff
@@ -8,12 +8,18 @@ source:
 
 requirements:
   build:
+    - cmake
+    - setuptools
+    - ninja # [win]
```
Comment (peterjc123) [inline conda/pytorch-cpu-0.3.1/meta.yaml]: Same here.

**Comment by Aidolii** [comment]
How close is this to getting merged?

**Comment by peterjc123** [comment]
@Aidolii At least we should merge this after bumping pytorch version.

**Comment by Aidolii** [comment]
Is this functional in the meantime? There doesn't seem to be any documentation for building it on Windows so I'm not entirely sure what to, I'm rather new to working with Anaconda.

I tried running one of the batch files and was met with a parade of errors. I'm curious what the correct step 1 is supposed to be and the proper working directory

**Comment by peterjc123** [comment]
@Aidolii You cannot directly use this script. Since v0.3.1 is not Windows ready, you should comment out the `git-rev` line and try again. Otherwise, you can use my version of 0.3.1, that is, https://github.com/peterjc123/pytorch/branch/v0.3.1_win.

**Comment by Aidolii** [comment]
Thank you, does your 0.3.1 branch work with CUDA 9.1?

**Comment by peterjc123** [comment]
@Aidolii I've tried that before and it should be okay.

**Comment by Aidolii** [comment]
Hah.. cuda91.bat... added 37 seconds ago.
